File: kmp/Pulsar/library/src/iosMain/kotlin/com/swmansion/pulsar/HapticEngineWrapper.kt
package com.swmansion.pulsar

import kotlinx.cinterop.ExperimentalForeignApi
import platform.CoreHaptics.CHHapticAdvancedPatternPlayerProtocol
import platform.CoreHaptics.CHHapticEngine
import platform.CoreHaptics.CHHapticEvent
import platform.CoreHaptics.CHHapticEventParameter
import platform.CoreHaptics.CHHapticEventParameterIDHapticIntensity
import platform.CoreHaptics.CHHapticEventParameterIDHapticSharpness
import platform.CoreHaptics.CHHapticEventTypeHapticContinuous
import platform.CoreHaptics.CHHapticPattern
import platform.CoreHaptics.CHHapticPatternPlayerProtocol
import platform.Foundation.NSLog
import platform.Foundation.NSNotification
import platform.Foundation.NSNotificationCenter
import platform.UIKit.UIApplication
import platform.UIKit.UIApplicationDidBecomeActiveNotification
import platform.UIKit.UIApplicationDidEnterBackgroundNotification
import platform.UIKit.UIApplicationWillEnterForegroundNotification
import platform.UIKit.UIApplicationWillResignActiveNotification
import platform.darwin.NSObject
import platform.objc.sel_registerName

@OptIn(ExperimentalForeignApi::class)
internal class IOSHapticEngineWrapper {
    private var engine: CHHapticEngine? = null
    private var initialized = false
    var isHapticsEnabled: Boolean = true
        private set
    private var isAppActive = true
    private val playerRegistry = mutableMapOf<Int, CHHapticPatternPlayerProtocol>()
    private val playerCreationOrder = mutableListOf<Int>()
    private var nextPlayerId = 0
    private var cachedRealtimePlayer: CHHapticAdvancedPatternPlayerProtocol? = null
    private val lifecycleObserver = IOSAppLifecycleObserver(this)

    init {
        if (!isHapticsSupported()) {
            log("Device does not support haptics")
        } else {
            startEngine()
            registerAppLifecycleObservers()
        }
    }

    fun enableHaptics(state: Boolean) {
        if (isHapticsEnabled == state) return
        isHapticsEnabled = state
        if (!state) {
            stopHaptics()
        } else if (canPlayHaptics() && !initialized) {
            startEngine()
        }
    }

    fun stopHaptics() {
        clearPlayerState(stopPlayers = true)
        runCatching { engine?.stopWithCompletionHandler(null) }
        initialized = false
    }

    fun shutDownEngine() {
        stopHaptics()
        engine = null
    }

    fun createPlayer(pattern: CHHapticPattern?): Int? {
        if (!canPlayHaptics()) return null
        startEngine()
        val player = buildPatternPlayer(pattern) ?: return null
        return registerPlayer(player)
    }

    fun getRealtimePlayer(): CHHapticAdvancedPatternPlayerProtocol? {
        if (!canPlayHaptics()) return null
        startEngine()
        cachedRealtimePlayer?.let { return it }
        return makeRealtimePlayer().also { cachedRealtimePlayer = it }
    }

    fun playPlayer(id: Int, pattern: CHHapticPattern? = null) {
        if (!canPlayHaptics()) return
        playerRegistry[id]?.let {
            startPlayer(it, "Error starting player")
            return
        }
        pattern ?: return
        startEngine()
        val player = buildPatternPlayer(pattern) ?: return
        registerRecreatedPlayer(player, id)
        startPlayer(player, "Error starting recreated player")
    }

    fun stopPlayer(id: Int) {
        playerRegistry[id]?.let { stopPlayer(it, "Error stopping player") }
    }

    fun removePlayer(id: Int) {
        runCatching { playerRegistry[id]?.stopAtTime(0.0, null) }
        playerRegistry.remove(id)
        playerCreationOrder.removeAll { it == id }
    }

    fun updatePlaybackAvailability(isActive: Boolean) {
        isAppActive = isActive
    }

    fun canPlayHaptics(): Boolean = isHapticsEnabled && isAppActive && isHapticsSupported()

    fun isHapticsSupported(): Boolean = CHHapticEngine.capabilitiesForHardware().supportsHaptics()

    private fun registerAppLifecycleObservers() {
        val center = NSNotificationCenter.defaultCenter
        center.addObserver(
            lifecycleObserver,
            selector = sel_registerName("appDidEnterBackground:"),
            name = UIApplicationDidEnterBackgroundNotification,
            `object` = null,
        )
        center.addObserver(
            lifecycleObserver,
            selector = sel_registerName("appWillEnterForeground:"),
            name = UIApplicationWillEnterForegroundNotification,
            `object` = null,
        )
        center.addObserver(
            lifecycleObserver,
            selector = sel_registerName("appDidBecomeActive:"),
            name = UIApplicationDidBecomeActiveNotification,
            `object` = null,
        )
        center.addObserver(
            lifecycleObserver,
            selector = sel_registerName("appWillResignActive:"),
            name = UIApplicationWillResignActiveNotification,
            `object` = null,
        )
    }

    fun appDidEnterBackground() {
        updatePlaybackAvailability(false)
        suspendHaptics()
    }

    fun appWillEnterForeground() {
        updatePlaybackAvailability(UIApplication.sharedApplication.applicationState.name == "UIApplicationStateActive")
    }

    fun appDidBecomeActive() {
        updatePlaybackAvailability(true)
        engine = null
        startEngine()
    }

    fun appWillResignActive() {
        updatePlaybackAvailability(false)
        suspendHaptics()
    }

    private fun suspendHaptics() {
        stopHaptics()
    }

    private fun startEngine() {
        if (initialized || !canPlayHaptics()) return
        runCatching {
            createEngineIfNeeded()
            engine?.startAndReturnError(null)
            initialized = true
        }.onFailure {
            initialized = false
            engine = null
            log("Error starting engine: ${it.message}")
        }
    }

    private fun createEngineIfNeeded() {
        if (engine != null) return
        engine = CHHapticEngine(andReturnError = null)
    }

    private fun buildPatternPlayer(pattern: CHHapticPattern?): CHHapticPatternPlayerProtocol? {
        val player = engine?.createPlayerWithPattern(
            pattern ?: CHHapticPattern(events = emptyList<Any>(), parameters = emptyList<Any>(), error = null),
            error = null,
        )
        if (player == null) log("Error making haptic player")
        return player
    }

    private fun makeRealtimePlayer(): CHHapticAdvancedPatternPlayerProtocol? {
        val event = CHHapticEvent(
            eventType = CHHapticEventTypeHapticContinuous,
            parameters = listOf(
                CHHapticEventParameter(CHHapticEventParameterIDHapticIntensity, 1.0f),
                CHHapticEventParameter(CHHapticEventParameterIDHapticSharpness, 0.0f),
            ),
            relativeTime = 0.0,
            duration = 100.0,
        )
        val pattern = CHHapticPattern(events = listOf(event), parameters = emptyList<Any>(), error = null)
        val player = engine?.createAdvancedPlayerWithPattern(pattern, error = null)
        if (player == null) log("Error creating realtime player")
        return player
    }

    private fun clearPlayerState(stopPlayers: Boolean) {
        if (stopPlayers) {
            playerRegistry.values.forEach { runCatching { it.stopAtTime(0.0, null) } }
            runCatching { cachedRealtimePlayer?.stopAtTime(0.0, null) }
        }
        playerRegistry.clear()
        playerCreationOrder.clear()
        cachedRealtimePlayer = null
    }

    private fun registerPlayer(player: CHHapticPatternPlayerProtocol): Int {
        evictOldestPlayerIfNeeded()
        val id = nextPlayerId++
        playerRegistry[id] = player
        playerCreationOrder += id
        return id
    }

    private fun registerRecreatedPlayer(player: CHHapticPatternPlayerProtocol, id: Int) {
        evictOldestPlayerIfNeeded()
        playerRegistry[id] = player
        playerCreationOrder += id
    }

    private fun evictOldestPlayerIfNeeded() {
        if (playerRegistry.size < 20) return
        val oldestId = playerCreationOrder.removeAt(0)
        runCatching { playerRegistry[oldestId]?.stopAtTime(0.0, null) }
        playerRegistry.remove(oldestId)
    }

    private fun startPlayer(player: CHHapticPatternPlayerProtocol, errorPrefix: String) {
        runCatching { player.startAtTime(0.0, null) }
            .onFailure { log("$errorPrefix: ${it.message}") }
    }

    private fun stopPlayer(player: CHHapticPatternPlayerProtocol, errorPrefix: String) {
        runCatching { player.stopAtTime(0.0, null) }
            .onFailure { log("$errorPrefix: ${it.message}") }
    }
}

internal class IOSAppLifecycleObserver(
    private val engine: IOSHapticEngineWrapper,
) : NSObject() {
    @Suppress("UNUSED_PARAMETER")
    fun appDidEnterBackground(notification: NSNotification) {
        engine.appDidEnterBackground()
    }

    @Suppress("UNUSED_PARAMETER")
    fun appWillEnterForeground(notification: NSNotification) {
        engine.appWillEnterForeground()
    }

    @Suppress("UNUSED_PARAMETER")
    fun appDidBecomeActive(notification: NSNotification) {
        engine.appDidBecomeActive()
    }

    @Suppress("UNUSED_PARAMETER")
    fun appWillResignActive(notification: NSNotification) {
        engine.appWillResignActive()
    }
}

internal fun log(message: String) {
    NSLog("Pulsar: $message")
}

File: flutter/pulsar/ios/Classes/HapticEngineWrapper.swift
import Foundation
import CoreHaptics
import SwiftUI

public class HapticEngineWrapper {
  private var engine: CHHapticEngine?
  private var initialized = false
  private(set) var isHapticsEnabled = true
  private var isAppActive: Bool
  private var playerRegistry: [Int: CHHapticPatternPlayer] = [:]
  private var playerCreationOrder: [Int] = []
  private let playerLimit = 20
  private var nextPlayerId = 0
  private var cachedRealtimePlayer: CHHapticAdvancedPatternPlayer?

  public init() {
    isAppActive = UIApplication.shared.applicationState == .active

    guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else {
      print("Error: Device doesn't support haptics")
      return
    }

    do {
      engine = try CHHapticEngine()
      setupEngineHandlers()
      startEngine()
      registerAppLifecycleObservers()
    } catch {
      print("Error starting engine: \(error.localizedDescription)")
    }
  }

  deinit {
    if !initialized { return }
    engine?.stop()
  }
}

// MARK: - Public API

public extension HapticEngineWrapper {

  func enableHaptics(_ state: Bool) {
    guard isHapticsEnabled != state else { return }

    isHapticsEnabled = state

    if !isHapticsEnabled {
      stopHaptics()
    } else if canPlayHaptics() && !initialized {
      startEngine()
    }
  }

  func stopHaptics() {
    clearPlayerState(stopPlayers: true)
    engine?.stop()
    initialized = false
  }

  func shutDownEngine() {
    stopHaptics()
    engine = nil
  }

  func createPlayer(pattern: CHHapticPattern?) -> Int? {
    guard canPlayHaptics() else { return nil }

    startEngine()
    guard let player = buildPatternPlayer(pattern: pattern) else { return nil }

    return registerPlayer(player)
  }

  func getRealtimePlayer() -> CHHapticAdvancedPatternPlayer? {
    guard canPlayHaptics() else { return nil }

    startEngine()
    if let player = cachedRealtimePlayer {
      return player
    }

    let player = makeRealtimePlayer()
    cachedRealtimePlayer = player
    return player
  }

  func playPlayer(id: Int, pattern: CHHapticPattern? = nil) {
    guard canPlayHaptics() else { return }

    if let player = playerRegistry[id] {
      startPlayer(player, errorPrefix: "Error starting player")
      return
    }

    guard let pattern else { return }

    startEngine()
    guard let player = buildPatternPlayer(pattern: pattern) else { return }

    registerRecreatedPlayer(player, for: id)
    startPlayer(player, errorPrefix: "Error starting recreated player")
  }

  func stopPlayer(id: Int) {
    guard let player = playerRegistry[id] else { return }
    stopPlayer(player, errorPrefix: "Error stopping player")
  }

  func removePlayer(id: Int) {
    if let player = playerRegistry[id] {
      try? player.stop(atTime: 0)
    }

    playerRegistry.removeValue(forKey: id)
    playerCreationOrder.removeAll { $0 == id }
  }
}

// MARK: - Internal State

extension HapticEngineWrapper {

  func updatePlaybackAvailability(for state: UIApplication.State) {
    isAppActive = state == .active
  }

  func canPlayHaptics() -> Bool {
    isHapticsEnabled && isAppActive && isHapticsSupported()
  }

  func isHapticsSupported() -> Bool {
    CHHapticEngine.capabilitiesForHardware().supportsHaptics
  }
}

// MARK: - App Lifecycle

private extension HapticEngineWrapper {

  func registerAppLifecycleObservers() {
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(appDidEnterBackground),
      name: UIApplication.didEnterBackgroundNotification,
      object: nil
    )
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(appWillEnterForeground),
      name: UIApplication.willEnterForegroundNotification,
      object: nil
    )
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(appDidBecomeActive),
      name: UIApplication.didBecomeActiveNotification,
      object: nil
    )
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(appWillResignActive),
      name: UIApplication.willResignActiveNotification,
      object: nil
    )
  }

  @objc func appDidEnterBackground() {
    updatePlaybackAvailability(for: .background)
    suspendHaptics()
  }

  @objc func appWillEnterForeground() {
    updatePlaybackAvailability(for: UIApplication.shared.applicationState)
  }

  @objc func appDidBecomeActive() {
    updatePlaybackAvailability(for: .active)
    engine = nil
    startEngine()
  }

  @objc func appWillResignActive() {
    updatePlaybackAvailability(for: .inactive)
    suspendHaptics()
  }

  func suspendHaptics() {
    stopHaptics()
  }
}

// MARK: - Engine Lifecycle

private extension HapticEngineWrapper {

  func startEngine() {
    guard !initialized, canPlayHaptics() else { return }

    do {
      try createEngineIfNeeded()
      try engine?.start()
      initialized = true
    } catch {
      print("Error starting engine: \(error.localizedDescription)")
      engine = nil
    }
  }

  func createEngineIfNeeded() throws {
    guard engine == nil else { return }

    engine = try CHHapticEngine()
    setupEngineHandlers()
  }

  func setupEngineHandlers() {
    engine?.stoppedHandler = { [weak self] _ in
      guard let self else { return }
      self.initialized = false
      self.clearPlayerState(stopPlayers: false)
    }

    engine?.resetHandler = { [weak self] in
      guard let self else { return }
      self.initialized = false
      self.engine = nil
    }
  }
}

// MARK: - Player Management

private extension HapticEngineWrapper {

  func buildPatternPlayer(pattern: CHHapticPattern?) -> CHHapticPatternPlayer? {
    do {
      return try engine?.makePlayer(with: pattern ?? CHHapticPattern(events: [], parameters: []))
    } catch {
      print("Error making pattern: \(error.localizedDescription)")
      return nil
    }
  }

  func makeRealtimePlayer() -> CHHapticAdvancedPatternPlayer? {
    let intensityParam = CHHapticEventParameter(parameterID: .hapticIntensity, value: 1)
    let sharpnessParam = CHHapticEventParameter(parameterID: .hapticSharpness, value: 0)
    let event = CHHapticEvent(
      eventType: .hapticContinuous,
      parameters: [intensityParam, sharpnessParam],
      relativeTime: 0,
      duration: 100
    )

    do {
      let pattern = try CHHapticPattern(events: [event], parameters: [])
      return try engine?.makeAdvancedPlayer(with: pattern)
    } catch {
      print("Error creating realtime player: \(error.localizedDescription)")
      return nil
    }
  }

  func clearPlayerState(stopPlayers: Bool) {
    if stopPlayers {
      stopAllRegisteredPlayers()
      try? cachedRealtimePlayer?.stop(atTime: 0)
    }

    playerRegistry.removeAll()
    playerCreationOrder.removeAll()
    cachedRealtimePlayer = nil
  }

  func stopAllRegisteredPlayers() {
    for player in playerRegistry.values {
      try? player.stop(atTime: 0)
    }
  }

  func registerPlayer(_ player: CHHapticPatternPlayer) -> Int {
    evictOldestPlayerIfNeeded()

    let id = nextPlayerId
    nextPlayerId += 1
    playerRegistry[id] = player
    playerCreationOrder.append(id)
    return id
  }

  func registerRecreatedPlayer(_ player: CHHapticPatternPlayer, for id: Int) {
    evictOldestPlayerIfNeeded()
    playerRegistry[id] = player
    playerCreationOrder.append(id)
  }

  func evictOldestPlayerIfNeeded() {
    guard playerRegistry.count >= playerLimit else { return }

    let oldestId = playerCreationOrder.removeFirst()
    if let player = playerRegistry[oldestId] {
      try? player.stop(atTime: 0)
    }
    playerRegistry.removeValue(forKey: oldestId)
  }
}

// MARK: - Player Playback

private extension HapticEngineWrapper {

  func startPlayer(_ player: CHHapticPatternPlayer, errorPrefix: String) {
    do {
      try player.start(atTime: 0)
    } catch {
      print("\(errorPrefix): \(error.localizedDescription)")
    }
  }

  func stopPlayer(_ player: CHHapticPatternPlayer, errorPrefix: String) {
    do {
      try player.stop(atTime: 0)
    } catch {
      print("\(errorPrefix): \(error.localizedDescription)")
    }
  }
}
